WordPress/gutenberg
# Data: Use deep equal instead of shallow equal for `useSelect`

## What?
This PR updates `useSelect()` to perform deep equality comparison instead of shallow equality comparison when updating values.

## Why?
While evaluating performance in the editor, I noticed that there are a bunch of rerenders for the same data. On a deeper inspection, I observed that some of them are with the exact same data. Then I noticed that when updating the value in `useSelect()`, we'll do a [shallow equality comparison](https://github.com/WordPress/gutenberg/blob/13c91846205c8bde00286aa6ada7abf3f575cf08/packages/data/src/components/use-select/index.js#L96) when attempting to preserve old values that are equal to the new ones, attempting to not trigger unwanted updates. This works fine, but only when comparing primitive values. When comparing objects that contain objects, even if they are equal, `isShallowEqual()` will intentionally return `false`:

```
const a = { foo: { bar: 1 } };
const b = { foo: { bar: 1 } };

isShallowEqual( a, b ); // false
isShallowEqual( b, a ); // false
```

That makes `useSelect()` inefficient when it retrieves more complex values that contain nested objects - it will always return new values, even when they are equal to the old ones. That causes unnecessary updates and re-renders as a consequence.

## How?
We're updating `useSelect()` to use `fastDeepEqual` instead of `@wordpress/is-shallow-equal`. `fastDeepEqual` is already widely used as a substitute for Lodash's `_.isEqual()`, so it's a safe bet.

We're introducing a new unit test and fixing an existing test to not expect a rerender for equal values. 

Feel free to try the new unit test against the `useSelect()` in `trunk` to see it failing.

## Testing Instructions
* Do some smoke testing to verify things still work well - typing, inserting blocks, moving blocks, transforming, etc.
* Verify all checks are green.

### Testing Instructions for Keyboard
None

## Screenshots or screencast 
None
